lib/storage/file.js

```diff
@@ -369,61 +364,76 @@ File.prototype.createReadStream = function(options) {
+              if (err) {
+                done(err);
+                return;
+              }
+
+              if (rangeRequest) {
+                // Range requests can't receive data integrity checks.
+                done(null, res);
+                return;
+              }
+
+              var failed = false;
+              var crcFail = true;
+              var md5Fail = true;
+
+              var hashes = {};
+              res.headers['x-goog-hash'].split(',').forEach(function(hash) {
+                var hashType = hash.split('=')[0];
+                hashes[hashType] = hash.substr(hash.indexOf('=') + 1);
+              });
+
+              var remoteMd5 = hashes.md5;
+              var remoteCrc = hashes.crc32c && hashes.crc32c.substr(4);
+
+              if (crc32c) {
+                crcFail =
+                  new Buffer([localCrcHash]).toString('base64') !== remoteCrc;
+                failed = crcFail;
+              }
+
+              if (md5) {
+                md5Fail = localMd5Hash.digest('base64') !== remoteMd5;
+                failed = md5Fail;
+              }
+
+              if (validation === 'all') {
+                failed = remoteMd5 ? md5Fail : crcFail;
+              }
+
+              if (failed) {
+                var mismatchError = new Error([
+                  'The downloaded data did not match the data from the server.',
+                  'To be sure the content is the same, you should download the',
+                  'file again.'
+                ].join(' '));
+                mismatchError.code = 'CONTENT_DOWNLOAD_MISMATCH';
+
+                done(mismatchError);
+              } else {
+                done(null, res);
+              }
             });
-
-            var remoteMd5 = hashes.md5;
-            var remoteCrc = hashes.crc32c && hashes.crc32c.substr(4);
-
-            if (crc32c) {
-              crcFail =
-                new Buffer([localCrc32cHash]).toString('base64') !== remoteCrc;
-              failed = crcFail;
-            }
-
-            if (md5) {
-              md5Fail = localMd5Hash.digest('base64') !== remoteMd5;
-              failed = md5Fail;
-            }
-
-            if (validation === 'all') {
-              failed = remoteMd5 ? md5Fail : crcFail;
-            }
-
-            if (failed) {
-              var error = new Error([
-                'The downloaded data did not match the data from the server.',
-                'To be sure the content is the same, you should download the',
-                'file again.'
-              ].join(' '));
-              error.code = 'CONTENT_DOWNLOAD_MISMATCH';
-
-              throughStream.emit('error', error);
-            } else {
-              throughStream.emit('complete', res);
-            }
-
-            throughStream.end();
           })
 
           .pipe(throughStream);
       }
     });
   }
+
+  // End the stream, first emitting an error or complete event.
+  function done(err, response) {
```

Comment:
New: use this function in a typical node callback style. If called with an error, emit an error on the stream, if not, emit complete with the response.